openedx_blockstore_relay/transfer_data.py

```diff
@@ -0,0 +1,257 @@
+"""
+Logic for transferring a subunit (aka XBlock, XModule) from Open edX to a Blockstore Bundle.
+"""
+from __future__ import absolute_import, print_function, unicode_literals
+
+import json
+import logging
+import os
+from xml.etree import ElementTree
+
+import requests
+from django.conf import settings
+from fs.memoryfs import MemoryFS
+from future.moves.urllib.error import HTTPError
+from future.moves.urllib.parse import urljoin
+
+LOG = logging.getLogger(__name__)
+
+
+def transfer_to_blockstore(block_key, bundle_uuid, asset_root_url=None):
+    """
+    Transfer the given block (and its children) to the given Blockstore bundle.
+    """
+    block_type = getattr(block_key, 'block_type', 'unknown')
+
+    if block_type == 'vertical':
+        transfer_obj = TransferUnit(block_key, bundle_uuid, asset_root_url)
+    elif block_type in ['sequential', 'chapter', 'course']:
+        # TODO: handle full descendent tree to support these block types.
+        raise NotImplementedError("Block type not supported: %s" % block_type)
+    else:
+        transfer_obj = TransferBlock(block_key, bundle_uuid, asset_root_url)
```

Comment:
Any reason why we can't use `TransferUnit` here too? For blocks without children the `for child in block.get_children():` loop should get skipped. Would be simpler to have generic code which descends the tree if there are child nodes and upload them one by one. It should automatically work for sequentials too. Something like:

```
def transfer_block_to_bundle():
  for child_block in block.get_children():
    transfer_child_block = TransferBlock(child_block)
    transfer_child_block.transfer_block_to_bundle()
  self.upload_assets()
  self.upload_olx()
```

Reply:
Yep, recursion is totally doable -- I was just holding off to limit the scope of this change 😄 

9e4e181 implements this, but it's not quite as clean as what you described, because:
* We don't want to update the `bundle.json` manifest file when transferring child blocks.
* We need to prevent the infinite recursion caused by Directed Acyclic Graphs.